### PR TITLE
Uninstall with versions

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
# Description:

Just like we support `gem install <name>:<requirement>`, I'd like to support `gem uninstall <name>:<requirement>`. Thoughts?
______________

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
@MSP-Greg This is failing under Windows on MRI 2.2, can you have a look?

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

> can you have a look?

Yes.  Don't use 2.2 much, but I've got both locally...

**Comment by deivid-rodriguez** [comment]
I'm actually wondering if it makes sense to drop support for it now that bundler has dropped it.

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

Both errors are related to Windows Ruby 2.2 not supporting symlinks.  There are two failing tests in `TestGemCommandsUninstallCommand`:

1\. `test_execute_two_version` - line 305 is currently
```
assert_equal msg, @ui.error.chomp
```
On Ruby 2.2, a warning about symlinks is 'issued', so the first line of `@ui.error` is the warning.  Replacing the line with the following passes.  Whether it should be wrapped in a conditional...
```
assert_equal msg, @ui.error.lines.last.chomp
```

2\. `test_execute_with_force_and_without_version_uninstalls_everything` - This one might take me a bit longer.  There's symlink code here and there, and in some places it just does a windows test, which probably isn't the best test.  I'll keep at it, but maybe tomorrow...

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

Got a fix for the failures.

`test_execute_with_force_and_without_version_uninstalls_everything` as it is currently in not actually removing an executable.  The issue is that `Gem::InstallerTestCase#setup` creates the gem with the executable (a-2), but `Gem::TestCase#util_make_gems` overwrites it.

This can be seen in the `assert_match "Removing executable", ui.output` statement, as adding it without the other statements shows that no executable is being removed.  This may be showing other things are amiss, but I don't think it's caused by the PR...

So, hope this helps.  Patch attached.

[pr_2466.patch.txt](https://github.com/rubygems/rubygems/files/2573992/pr_2466.patch.txt)

**Comment by deivid-rodriguez** [comment]
Thank you @MSP-Greg! Feel free to open a PR to my fork so you can keep attribution!

**Comment by MSP-Greg** [comment]
> open a PR to my fork

Done.

**Comment by deivid-rodriguez** [comment]
Thanks so much @MSP-Greg! I merged your PR into this branch. I should've problably squashed it to remove the merge commit, I'll do it manually once CI passes.

File: lib/rubygems/commands/uninstall_command.rb [hunk 1]
```diff
@@ -114,7 +114,18 @@ def usage # :nodoc:
     "#{program_name} GEMNAME [GEMNAME ...]"
   end
 
+  def check_version # :nodoc:
+    if options[:version] != Gem::Requirement.default and
+         get_all_gem_names.size > 1 then
+      alert_error "Can't use --version with multiple gems. You can specify multiple gems with" \
+                  " version requirments using `gem install 'my_gem:1.0.0' 'my_other_gem:~>2.0.0'`"
```
Comment (bronzdoc) [inline lib/rubygems/commands/uninstall_command.rb]: "requirements"

File: test/rubygems/test_gem_commands_uninstall_command.rb [hunk 2]
```diff
@@ -261,6 +291,25 @@ def test_handle_options_vendor
...
+    use_ui @ui do
+      e = assert_raises Gem::MockGemUi::TermError do
+        @cmd.execute
+      end
+
+      assert_equal 1, e.exit_code
+    end
+
+    msg = "ERROR:  Can't use --version with multiple gems. You can specify multiple gems with" \
+      " version requirments using `gem install 'my_gem:1.0.0' 'my_other_gem:~>2.0.0'`"
```
Comment (bronzdoc) [inline test/rubygems/test_gem_commands_uninstall_command.rb]: "requirements"

Comment (MSP-Greg) [reply test/rubygems/test_gem_commands_uninstall_command.rb]: @bronzdoc

Sorry, I was in the test file and missed that.  Two items:

1. @deivid-rodriguez just copied it from here, where the typo also exists:

https://github.com/rubygems/rubygems/blob/4dc3d9d5d1a92f5b1b056716900057b016d9ea1b/lib/rubygems/commands/install_command.rb#L144-L145

2. In the two uninstall files here, should "gem install' be changed to "gem uninstall" in the 'messages'?

Comment (MSP-Greg) [reply test/rubygems/test_gem_commands_uninstall_command.rb]: @bronzdoc

`install_command.rb` is modified in #2452, should I fix the typo there?

**Comment by deivid-rodriguez** [comment]
@MSP-Greg I reverted your patch since it's no longer necessary because of #2487. I blindly reverted the whole commit, but if there were specific changes there aimed to improve the tests in general and not just get them passing, just let me know and I can add those back!

**Comment by deivid-rodriguez** [comment]
Thank you, I'll add the commit back then!

**Comment by deivid-rodriguez** [comment]
This PR should be ready to go now!

**Comment by bronzdoc** [comment]
@bundlerbot r+

**Comment by yizikkn** [comment]
Styling kingdom

**Pull request merged**